### PR TITLE
Allow paths in filename

### LINKED ISSUE #5: tass <action> --file should take a file path
To conform to general UNIX-y patterns. Running the tool from within the config/clusters directory is kind of counterintuitive.

### PR DESCRIPTION
**Pull request by atayarani**
@Tapjoy/eng-group-ops @marcuswalser  @StabbyCutyou 

Fixes #5

### CODE REVIEW

File: lib/tapjoy/autoscaling_bootstrap.rb [hunk 1]
```diff
@@ -154,13 +154,22 @@ def confirm_config(keypair:, zones:, security_groups:, instance_type:,
...
+      def configure_environment(filename, env=nil, config_dir=nil)
+        if filename.include?(File::SEPARATOR)
+          facet_file    = filename
+          config_dir   = "#{File.expand_path('../../..', facet_file)}"
+        else
+          facet_file    = "#{config_dir}/config/clusters/#{filename}"
+        end
+
+        common_path   = "#{config_dir}/config/common/"
+        defaults_hash = self.load_yaml("#{common_path}/defaults.yaml")
```
Comment (StabbyCutyou) [inline lib/tapjoy/autoscaling_bootstrap.rb]: A more idiomatic way to build file paths would be to use File.join("part1","part2","part3"), which would result in "part1/part2/part3" (the first part needs a leading slash if you intend for it to lead with one, which File.expand_path does).

That's not really a blocking concern, but just something to keep in mind

**Comment by atayarani** [comment]
@StabbyCutyou I've changed it in 5069c05

Is that what you had in mind?

File: lib/tapjoy/autoscaling_bootstrap.rb [hunk 2]
```diff
@@ -154,13 +154,22 @@ def confirm_config(keypair:, zones:, security_groups:, instance_type:,
...
 
       # configure environment
-      def configure_environment(filename, env=nil, config_dir)
-        defaults_hash = self.load_yaml("#{config_dir}/config/common/defaults.yaml")
-        facet_hash    = self.load_yaml("#{config_dir}/config/clusters/#{filename}")
+
+      def configure_environment(filename, env=nil, config_dir=nil)
+        if filename.include?(File::SEPARATOR)
+          facet_file    = filename
+          config_dir   = "#{File.expand_path('../../..', facet_file)}"
```
Comment (StabbyCutyou) [inline lib/tapjoy/autoscaling_bootstrap.rb]: You don't need to string interpolate this, expand_path returns a string

File: lib/tapjoy/autoscaling_bootstrap.rb [hunk 3]
```diff
@@ -154,13 +154,22 @@ def confirm_config(keypair:, zones:, security_groups:, instance_type:,
...
+        else
+          facet_file    = File.join(config_dir, 'config', 'clusters', filename)
```
Comment (StabbyCutyou) [inline lib/tapjoy/autoscaling_bootstrap.rb]: if config dir is nil, it'll cause some issues (plus you would lose the leading slash on 'config' from your prior example, if it didn't already have one)

you could likely do something like

```ruby
File.join(config_dir || "/",'config','clusters',filename)
```

Which would fallback to using a vanilla leading slash only if config_dir is nil

Comment (StabbyCutyou) [inline lib/tapjoy/autoscaling_bootstrap.rb]: To note, the problem is that File.join doesn't like nils

File: lib/tapjoy/autoscaling_bootstrap.rb [hunk 4]
```diff
@@ -154,13 +154,22 @@ def confirm_config(keypair:, zones:, security_groups:, instance_type:,
...
+        end
+
+        common_path   = File.join(config_dir, 'config', 'common')
+        defaults_hash = self.load_yaml("#{common_path}/defaults.yaml")
+        facet_hash    = self.load_yaml(facet_file)
         env         ||= facet_hash[:environment]
         env         ||= defaults_hash[:environment]
         Tapjoy::AutoscalingBootstrap.is_valid_env?(config_dir, env)
-        env_hash      = self.load_yaml("#{config_dir}/config/common/#{env}.yaml")
+        env_hash      = self.load_yaml("#{common_path}/#{env}.yaml")
```
Comment (StabbyCutyou) [inline lib/tapjoy/autoscaling_bootstrap.rb]: File.join here too

**Comment by StabbyCutyou** [comment]
Yup!

Couple of edge cases to consider, but otherwise thats a little more in line with how you should think about constructing paths in Ruby.

**Comment by atayarani** [comment]
Okay, I removed the nil and added more File.join in 54d8fbf

**Comment by StabbyCutyou** [comment]
Looks fine from a ruby perspective, can't claim it works :)

**Comment by marcuswalser** [comment]
When building the gem, I get the following error:

WARNING:  pessimistic dependency on hashdiff (~> 0.2.2) may be overly strict
  if hashdiff is semantically versioned, use:
    add_runtime_dependency 'hashdiff', '~> 0.2', '>= 0.2.2'

**Comment by StabbyCutyou** [comment]
It's a warning, not a hard error.

It's telling you that you might be overly restrictive on your dependencies
for your own good, but it's only an issue if you can't use the gem in the
project you want to.

**Comment by atayarani** [comment]
Also, I just updated to it to be less strict.

**Comment by marcuswalser** [comment]
create, update and audit still work, gem installs fine. LGTM

**Pull request merged**